library/alloc/src/collections/vec_deque/iter.rs

```diff
@@ -13,38 +15,53 @@ use super::{count, wrap_index, RingSlices};
 /// [`iter`]: super::VecDeque::iter
 #[stable(feature = "rust1", since = "1.0.0")]
 pub struct Iter<'a, T: 'a> {
-    ring: &'a [MaybeUninit<T>],
+    ring: NonNull<[T]>,
     tail: usize,
     head: usize,
+    _marker: PhantomData<&'a T>,
 }
 
 impl<'a, T> Iter<'a, T> {
     pub(super) fn new(ring: &'a [MaybeUninit<T>], tail: usize, head: usize) -> Self {
-        Iter { ring, tail, head }
+        Iter {
+            ring: unsafe { NonNull::new_unchecked(ring as *const [MaybeUninit<T>] as *mut _) },
+            tail,
+            head,
+            _marker: PhantomData,
+        }
+    }
+
+    unsafe fn ring(&self) -> &'a [MaybeUninit<T>] {
```

Comment:
That's a very long-lived reference, is that needed?

Reply:
As far as I can tell, we need to get a `&'a T` by some means in order to implement all the iterator methods, so we're stuck with this hazard.

Let's see if I can do something to just `Drain` to fix the problem there...

Reply:
The second commit in this PR now reverts the first one and modifies just the `Drain` struct. Do you think that approach is more satisfying?

Reply:
I haven't checked the details but I do think that reusing `Iter` was kind of a hack so this seems cleaner to me indeed.

Let's see what the libs reviewer says.

Reply:
Seems fine to me.